### PR TITLE
Destructured exports no unused modules

### PR DESCRIPTION
**Pull request by KristjanTammekivi**
Fixes https://github.com/benmosher/eslint-plugin-import/issues/2037

### CODE REVIEW

File: tests/src/rules/no-unused-modules.js [hunk 1]
```diff
@@ -663,6 +663,33 @@ describe('test behaviour for new file', () => {
   });
 });
 
+describe('test behaviour for destructured exports', () => {
+  before(() => {
```
Comment (ljharb) [inline tests/src/rules/no-unused-modules.js:667]: Please always use beforeEach/afterEach; before/after is “all” semantics, which leads to test pollution.

File: tests/src/rules/no-unused-modules.js [hunk 2]
```diff
@@ -663,6 +663,33 @@ describe('test behaviour for new file', () => {
   });
 });
 
+describe('test behaviour for destructured exports', () => {
```
Comment (ljharb) [inline tests/src/rules/no-unused-modules.js:666]:
```suggestion
describe('test behavior for destructured exports', () => {
```

**Review by ljharb — CHANGES_REQUESTED** [review]
Named exports aren’t destructuring; please update the terminology :-)

**Comment by KristjanTammekivi** [comment]
I don't see how `export const { a } = { a: 5 };` is not destructuring? In MDN there's an [example](https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export) of destructured assignments
```js
// Exporting destructured assignments with renaming
export const { name1, name2: bar } = o;
```

**Comment by ljharb** [comment]
Ohh, i misunderstood! Thanks.

File: tests/src/rules/no-unused-modules.js [hunk 3]
```diff
@@ -663,7 +663,34 @@ describe('test behaviour for new file', () => {
   });
 });
 
-describe('test behaviour for new file', () => {
+describe('test behavior for destructured exports', () => {
+  beforeEach(() => {
+    fs.writeFileSync(testFilePath('./no-unused-modules/file-added-3.js'), '', { encoding: 'utf8' });
```
Comment (ljharb) [inline tests/src/rules/no-unused-modules.js:668]: Is there a reason not to make this just be a checked-in empty file? We have a number of fixtures already.

Comment (KristjanTammekivi) [reply tests/src/rules/no-unused-modules.js]: There's the reason that I was just hacking it together copying and pasting as a proof of concept :)
I sort of fixed it now but I couldn't figure out why, even though I reset the empty_file in afterEach, I had to change the invalid case to a different variable name to make it fail.

File: tests/src/rules/no-unused-modules.js [hunk 4]
```diff
@@ -663,7 +663,35 @@ describe('test behaviour for new file', () => {
   });
 });
 
-describe('test behaviour for new file', () => {
+describe('test behavior for destructured exports', () => {
+  beforeEach(() => {
+    fs.writeFileSync(testFilePath('./no-unused-modules/file-destructured.js'), '', { encoding: 'utf8' });
+  });
```
Comment (ljharb) [inline tests/src/rules/no-unused-modules.js:669]: my hope is to avoid any fs operations here - can these not be committed, with their contents?

Comment (KristjanTammekivi) [reply tests/src/rules/no-unused-modules.js:669]: Uhh, sure just a moment, not sure why You prefer committed files though

Comment (ljharb) [reply tests/src/rules/no-unused-modules.js:669]: It's more deterministic, doesn't require any sort of permissions to run tests, and is much easier to debug.

**Review by ljharb — APPROVED** [review]
Thanks, seems good

**Comment by ljharb** [comment]
(Note that merging this is blocked on #1986)

**Comment by ljharb** [comment]
This was actually fixed by #1997, but I've rebased this to keep the test.

**Comment by emreboga** [comment]
Hey @ljharb, I don't think this was fully fixed by #1997 as it only addresses object type destucturing, not including array types. You can see it in the code:
https://github.com/import-js/eslint-plugin-import/pull/1997/files#diff-7d2908b010ec724c0f3a0d7615cb91f4673d765d52c6ba6607501e23470f2966R85

I'm seeing this in the code with the version 2.28.1, which has the fix with #1997. Here is a code chunk I have:
```
export const [authUserLoading, authUserError] = createLoadingAndErrorSelectors(
  AUTH_USER
);
```
and here is the error I get:
```
error  exported declaration 'undefined' not used within other modules  import/no-unused-modules
```

**Comment by ljharb** [comment]
@emreboga interesting - would you mind filing a new issue about that one?

**Comment by emreboga** [comment]
@ljharb Sure! Please see https://github.com/import-js/eslint-plugin-import/issues/2930

**Pull request merged**